modules/setting/queue.go

```diff
@@ -0,0 +1,214 @@
+	q.BlockTimeout = sec.Key("BLOCK_TIMEOUT").MustDuration(Queue.BlockTimeout)
+	q.BoostTimeout = sec.Key("BOOST_TIMEOUT").MustDuration(Queue.BoostTimeout)
+	q.BoostWorkers = sec.Key("BOOST_WORKERS").MustInt(Queue.BoostWorkers)
+	q.QueueName = sec.Key("QUEUE_NAME").MustString(Queue.QueueName)
+
+	q.Network, q.Addresses, q.Password, q.DBIndex, _ = ParseQueueConnStr(q.ConnectionString)
+	return q
+}
+
+// NewQueueService sets up the default settings for Queues
+// This is exported for tests to be able to use the queue
+func NewQueueService() {
+	sec := Cfg.Section("queue")
+	Queue.DataDir = sec.Key("DATADIR").MustString("queues/")
+	if !path.IsAbs(Queue.DataDir) {
+		Queue.DataDir = path.Join(AppDataPath, Queue.DataDir)
+	}
+	Queue.Length = sec.Key("LENGTH").MustInt(20)
+	Queue.BatchLength = sec.Key("BATCH_LENGTH").MustInt(20)
+	Queue.ConnectionString = sec.Key("CONN_STR").MustString(path.Join(AppDataPath, ""))
+	validTypes := queue.RegisteredTypesAsString()
+	Queue.Type = sec.Key("TYPE").In(string(queue.PersistableChannelQueueType), validTypes)
+	Queue.Network, Queue.Addresses, Queue.Password, Queue.DBIndex, _ = ParseQueueConnStr(Queue.ConnectionString)
+	Queue.WrapIfNecessary = sec.Key("WRAP_IF_NECESSARY").MustBool(true)
+	Queue.MaxAttempts = sec.Key("MAX_ATTEMPTS").MustInt(10)
+	Queue.Timeout = sec.Key("TIMEOUT").MustDuration(GracefulHammerTime + 30*time.Second)
+	Queue.Workers = sec.Key("WORKERS").MustInt(1)
+	Queue.MaxWorkers = sec.Key("MAX_WORKERS").MustInt(10)
+	Queue.BlockTimeout = sec.Key("BLOCK_TIMEOUT").MustDuration(1 * time.Second)
+	Queue.BoostTimeout = sec.Key("BOOST_TIMEOUT").MustDuration(5 * time.Minute)
+	Queue.BoostWorkers = sec.Key("BOOST_WORKERS").MustInt(5)
+	Queue.QueueName = sec.Key("QUEUE_NAME").MustString(Queue.QueueName)
+
+	hasWorkers := false
+	for _, key := range Cfg.Section("queue.notification").Keys() {
+		if key.Name() == "WORKERS" {
+			hasWorkers = true
+			break
+		}
+	}
+	if !hasWorkers {
+		Cfg.Section("queue.notification").Key("WORKERS").SetValue("5")
+	}
+
+	// Now handle the old issue_indexer configuration
+	section := Cfg.Section("queue.issue_indexer")
+	issueIndexerSectionMap := map[string]string{}
+	for _, key := range section.Keys() {
+		issueIndexerSectionMap[key.Name()] = key.Value()
+	}
+	if _, ok := issueIndexerSectionMap["TYPE"]; !ok {
+		switch Indexer.IssueQueueType {
+		case LevelQueueType:
+			section.Key("TYPE").SetValue("level")
+		case ChannelQueueType:
+			section.Key("TYPE").SetValue("persistable-channel")
+		case RedisQueueType:
+			section.Key("TYPE").SetValue("redis")
+		default:
+			log.Fatal("Unsupported indexer queue type: %v",
+				Indexer.IssueQueueType)
+		}
+	}
+	if _, ok := issueIndexerSectionMap["LENGTH"]; !ok {
+		section.Key("LENGTH").SetValue(fmt.Sprintf("%d", Indexer.UpdateQueueLength))
+	}
+	if _, ok := issueIndexerSectionMap["BATCH_LENGTH"]; !ok {
+		section.Key("BATCH_LENGTH").SetValue(fmt.Sprintf("%d", Indexer.IssueQueueBatchNumber))
```

Comment:
What will the pusher do if the queue is full? Will it block, will it drop requests? (e.g. in the initial indexer population, a repo migration, etc).

Reply:
You should be commenting on the Length setting - Batch_length is about the size of the array sent to the handler of the queue.

However, if the queue contains LENGTH elements it blocks - hence the pool worker boost ability.

Reply:
I ask because IMHO I think it _should_ block. Let's take indexing as example. In the initial index population, if there are too many files to process, the indexing will not get faster by adding workers (i.e. boost): I/O will be at peak demand, memory consumption will skyrocket, the whole system will slow down... even start swapping out... However, if the queue blocks (which is a natural thing to do for a queue), it will make the data provider (repository loop) to wait until some of the work has been cleared out. On the other hand, if releasing the provider as quick as possible is critical, it's preferrable to let the queue grow indefinitely and have only the right amount of workers.

Reply:
OK, so it is dependent on the underlying queue type where it blocks.

In the case of the level and redis queues the caller will not block - it's going to block at the reader from the db and it is there that handles the boost functionality.

In the case of the channel queues it will block at caller level. Persistent Channel queues are basically Channel queues except on initialisation when they read from the disk or at shutdown when they write to the disk.

Whether you get a boost or not is dependent on the pool settings but by default you shouldn't end up with more than 10 workers on any particular queue - with a boost of 5 workers if you block for a 1s and a further boost of 4 if you block for 2s after blocking for 1s.